pkg/didcomm/protocol/messagepickup/service.go

```diff
@@ -0,0 +1,612 @@
+/*
+Copyright Scoir Inc. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package messagepickup
+
+import (
+	"encoding/json"
+	"fmt"
+	"math"
+	"sync"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/pkg/errors"
+
+	"github.com/hyperledger/aries-framework-go/pkg/common/log"
+	"github.com/hyperledger/aries-framework-go/pkg/didcomm/common/model"
+	"github.com/hyperledger/aries-framework-go/pkg/didcomm/common/service"
+	commtransport "github.com/hyperledger/aries-framework-go/pkg/didcomm/common/transport"
+	"github.com/hyperledger/aries-framework-go/pkg/didcomm/dispatcher"
+	"github.com/hyperledger/aries-framework-go/pkg/didcomm/protocol/decorator"
+	"github.com/hyperledger/aries-framework-go/pkg/didcomm/transport"
+	"github.com/hyperledger/aries-framework-go/pkg/framework/aries/api"
+	"github.com/hyperledger/aries-framework-go/pkg/storage"
+	"github.com/hyperledger/aries-framework-go/pkg/store/connection"
+)
+
+const (
+	// MessagePickup defines the protocol name
+	MessagePickup = "messagepickup"
+	// Spec defines the protocol spec
+	Spec = "https://didcomm.org/messagepickup/1.0/"
+	// StatusMsgType defines the protocol propose-credential message type.
+	StatusMsgType = Spec + "status"
+	// StatusRequestMsgType defines the protocol propose-credential message type.
+	StatusRequestMsgType = Spec + "status-request"
+	// BatchPickupMsgType defines the protocol offer-credential message type.
+	BatchPickupMsgType = Spec + "batch-pickup"
+	// BatchMsgType defines the protocol offer-credential message type.
+	BatchMsgType = Spec + "batch"
+	// NoopMsgType defines the protocol request-credential message type.
+	NoopMsgType = Spec + "noop"
+)
+
+const (
+	updateTimeout = 50 * time.Second
+
+	// Namespace is namespace of messagepickup store name
+	Namespace = "mailbox"
+)
+
+// ErrConnectionNotFound connection not found error
+var ErrConnectionNotFound = errors.New("connection not found")
+var logger = log.New("aries-framework/messagepickup")
+
+type provider interface {
+	OutboundDispatcher() dispatcher.Outbound
+	StorageProvider() storage.Provider
+	TransientStorageProvider() storage.Provider
+}
+
+type connections interface {
+	GetConnectionRecord(string) (*connection.Record, error)
+}
+
+// Service for the messagepickup protocol
+type Service struct {
+	service.Action
+	service.Message
+	connectionLookup connections
+	outbound         dispatcher.Outbound
+	msgStore         storage.Store
+	packager         commtransport.Packager
+	msgHandler       transport.InboundMessageHandler
+	batchMap         map[string]chan Batch
+	batchMapLock     sync.RWMutex
+	statusMap        map[string]chan Status
+	statusMapLock    sync.RWMutex
+	inboxMap         map[string]*lockBox
+	inboxMapLock     sync.RWMutex
+}
+
+type lockBox struct {
+	mux sync.RWMutex
+}
```

Comment:
If you need lock by key please check  the following links:
https://github.com/kubernetes/utils/blob/master/keymutex/hashed.go
https://github.com/im7mortal/kmutex